aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -298,6 +298,59 @@ Tensor& sigmoid_out(Tensor& result, const Tensor& self) { return unary_op_impl_o
 Tensor sigmoid(const Tensor& self) { return unary_op_impl(self, at::sigmoid_out);  }
 Tensor& sigmoid_(Tensor& self) { return unary_op_impl_(self, at::sigmoid_out);  }
 
+Tensor& heaviside_tensor_out(Tensor& result, const Tensor& self, const Tensor& val) {
+  TORCH_CHECK(!self.is_complex() && !result.is_complex() && !val.is_complex(),
+    "heaviside is not implemented for complex tensors.");
+  TORCH_CHECK(result.is_floating_point(),
+    "heaviside is only implemented for floating point types output.");
+
+  result = result.to(kFloat);
```

Comment:
This seems very interesting that, In NumPy, no matter what dtype of input parameters is, we will always get a `float64` result. So I imitated numpy in my function and let its results keeping in `float32`, which is the default float-point dtype in pytorch. Does this make sense?

Reply:
Thank you for being so thorough in your review, @Kiyosora! I find it hard to believe this behavior is intended in NumPy, however. And looking at the heaviside PR it doesn't appear. See this final comment:

https://github.com/numpy/numpy/pull/8795#issuecomment-314276058

cc @rgommers as an fyi.

I would implement standard PyTorch binary op type promotion (which you'll get automatically if you use the binary_op function to construct your TensorIterator).